### ISSUE TITLE
Proposal: Documentation.

### ISSUE THREAD

**Issue by Kesanov** [ethereum/moon-lang#25]
So I think every moon file should like this:

```haskell
-- concatenates two lists   -- part of doc string
this : -- right now will behave as special version of doc string
  List = a => b => 
    ((a -> b -> b) -> b -> List a b)
  (List a b)
  -> (List a c)
  -> (List a c)

this =
  xs => -- list -- part of doc string
  ys => -- list -- part of doc string
    val => end => (xs val (ys val end))

-- EXAMPLES --

val = ...
end = ...

(this end end) == end
(this end (val 1 end)) == (val 1 end)
(this (val 1 end) end) == (val 1 end)
(this (val 1 end) (val 2 end)) == (val 1 (val 2 end))
```

The type signature should behave like a comment right now and later can be replaced with real type system.

REPL then should have available three commands: `:type` `:documentation` `:example` with obvious consequences.

I have already rewritten whole list library with this style, yet this requires some changes on the compiler part `->`, `:` and optionally `==`, which I might do on my own if you do not have time.

**Comment by VictorTaelin** [comment]
I'm more in favor of having the types annotated together with λ-variables, but that is a minor issue. The real problem with type system is that we simply don't have one that is small, consistent, practical and expressive enough to express induction, pattern-matching and prove simple things such as `1 != 0`. 

Morte (the calculus of constructions), for one, can't express induction, `1 != 0` and pattern-matching. The calculus of inductive constructions (what Coq is based on) and similar don't really qualify as "small" IMO, because of GADTs, which are very complex. I've been discussing about self types on the Morte repository, and there's a recent paper on `ιλp2`, things like those might be the solution, but they're not 100% convincing. Another way would be to use EAL as the base logic, which gives us all we need, but might not be very practical.

So, in short, there is no single, clear answer on the right way to do it. I'm left with no options other than waiting for further insights. If you have a one, please let me know.

**Comment by VictorTaelin** [comment]
You said you have this working already, though, so what do you mean? What kind of type system did you implement? I'd like to see it.

**Comment by Kesanov** [comment]
Oh, that was a misunderstanding. I am aware moon does not have type system, and will not have for quite a while. But types are very helpful in understanding how something works. Therefore I want to make types signatures part of documentation (they wont be type checked).

Then `:` and `->` are just aliases for `--` with behaviour of `=` and `=>` respectively.

**Comment by Kesanov** [comment]
This proposal is mainly about how should moon documentation look like.

**Comment by Kesanov** [comment]
> I'm more in favor of having the types annotated together with λ-variables

Do you mean:

```haskell
this =
  xs : (List a b) => -- list -- part of doc string
  ys : (List a c) => -- list -- part of doc string
    val => end => (xs val (ys val end))
```

But where do you put the return type and the definition of `List = a => b => ((a -> b -> b) -> b -> List a b)`.

**Comment by Kesanov** [comment]
It will be better if we continue our discussion in the PR.

**Comment by VictorTaelin** [comment]
Oh, woops. OK.

**Comment by VictorTaelin** [comment]
> But where do you put the return type and the definition of List = a => b => ((a -> b -> b) -> b -> List a b).  

Nowhere, as you can always infer it, but you can optionally use `the` (the function from Idris, `the : ∀ (t : *) . ∀ (a : t) . a`) to annotate the returned value (or any sub expression).
